src/Vault.sol

```diff
@@ -0,0 +1,175 @@
+    }
+    
+    ///@notice Users borrow debt asset calculated based on collateralization level and their deposits 
+    ///@dev See getMaxDebt() for colleteralization calculation
+    ///@param debtAmount Amount of debt asset to borrow
+    function borrow(uint debtAmount) external {             
+        uint collateralRequired = getCollateralRequired(debtAmount);
+        require(collateralRequired < deposits[msg.sender], "Insufficient collateral!");
+
+        debts[msg.sender] += debtAmount;
+        bool sent = debt.transfer(msg.sender, debtAmount);
+        require(sent, "Borrow failed!");       
+        emit Borrow(address(debt), msg.sender, debtAmount);
+    }
+
+    ///@notice Users repay their debt, in debt asset terms
+    ///@dev This covers partial and full repayment
+    ///@param debtAmount Amount of debt asset to repay
+    function repay(uint debtAmount) external {
+        debts[msg.sender] -= debtAmount;
+
+        bool sent = debt.transferFrom(msg.sender, address(this), debtAmount);
+        require(sent, "Repayment failed!");       
+        emit Repay(address(debt), msg.sender, debtAmount);   
+    }
+
+
+    ///@notice Users withdraw their deposited collateral
+    ///@dev This covers partial and full withdrawal; checks for spare capacity before initiating withdrawal
+    ///@param collateralAmount Amount of collateral asset to withdraw
+    function withdraw(uint collateralAmount) external {       
+        uint userDebt = debts[msg.sender];                          // caching this saves an SLOAD
+        uint userDeposit = deposits[msg.sender];                   // caching this saves an SLOAD
+
+        if (userDebt > 0){
+            uint collateralRequired = getCollateralRequired(userDebt);
+            uint spareDeposit = userDeposit - collateralRequired;
+            require(collateralAmount < spareDeposit, "Collateral unavailable!");
+        }
+        
+        deposits[msg.sender] -= collateralAmount;
+        bool sent = collateral.transfer(msg.sender, collateralAmount);
+        require(sent, "Withdraw failed!");
+        emit Withdraw(address(collateral), msg.sender, collateralAmount);            
+    }
+
+    ///@notice Price is returned as an integer extending over its decimal places
+    ///@dev Calculates collateral required to support existing debt position at current market prices 
+    ///@param debtAmount Amount of debt asset to support
+    /*Note: Example calculation using collateralLevel (e.g. DAI/WETH):
+    *       baseCollateralRequired = 1 WETH (1e18), collateralLevel = 80% (8e17), scalarFactor = 10**18  (DAI/WETH is 18dp)
+    *       1 WETH (1e18) * 80% (8e17) / 10**18 = 8e(18+17) / 10**18 = 8e17  => 80% of 1 WETH
+    *       Since we want to have a buffer, we will DIVIDE by collateralLevel: 1/0.8 = 1.25 (20% buffer, 80% collateralization level)
+    */
+    function getCollateralRequired(uint debtAmount) public view returns(uint) {
```

Comment:
Maybe changing this function for something like `function isCollateralized(address user) view returns (book)` or `collateralizer

Reply:
Wouldn't it be necessary for the end-user to know how much collateral would be required for a certain debt amount, before committing to said debt amount? Like a preview assessment of sorts before pulling the trigger.

Reply:
That's a good point, and is something that we did for ERC4626 with all those `max*` functions.

The answer to your question is "it depends". If your end user is going to use a frontend, that frontend can probably calculate those numbers.

Less code in a smart contract is generally a good thing, but when we are talking about view functions that have little risk.

This kind of helper view functions become useful with contract integration.  We implemented [some](https://github.com/yieldprotocol/vault-v2/blob/0b0926c3f7c0f06677a450a52a36352334b71cc5/contracts/Cauldron.sol#L271) in our Collateralized Debt Engine, and they ended up being quite useful.

Long story short, let's simplify the code logic as much as we can (one single function to judge collateralization) but be quite open to introduce zero-risk helper functions that give information that might be useful to integrators.

Reply:
So I guess that you should do a `isCollateralized` function for internal use, just because of how much it simplifies your core logic, and then you can implement `minCollateral` and `maxDebt` external functions, because they will be useful.

Reply:
I'm really unsure how I could be simplifying it into isCollateralized as you have suggested. I seem to come back to the same functionality out of necessity given that `borrow()` relies on `getCollateralRequired()` as a collateralization check,

```
    function borrow(uint debtAmount) external {      
        uint collateralRequired = getCollateralRequired(debtAmount);
        uint availableCollateral = deposits[msg.sender] - getCollateralRequired(debts[msg.sender]);
        require(collateralRequired < availableCollateral, "Insufficient collateral!");

        debts[msg.sender] += debtAmount;
        bool sent = debt.transfer(msg.sender, debtAmount);
        require(sent, "Borrow failed!");       
        emit Borrow(address(debt), msg.sender, debtAmount);
    }
```

```
    function getCollateralRequired(uint debtAmount) public view returns(uint) {
        (,int price,,,) = priceFeed.latestRoundData();
        uint collateralRequired = debtAmount * uint(price) / scalarFactor;
        return collateralRequired = scaleDecimals(collateralRequired, collateralDecimals, debtDecimals);
    }
```